api/client.go

```diff
@@ -57,13 +61,13 @@ func (cfg *Config) roundTripper() http.RoundTripper {
 	return cfg.RoundTripper
 }
 
-func (cfg *Config) client() http.Client {
+func (cfg *Config) client() HttpClient {
```

Comment:
```suggestion
func (cfg *Config) client() *http.Client {
```
Let's return to the concrete type. So we don't restrict the caller and ourselves from introducing any breaking changes to the interfaces.

https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Reply:
Hi @kakkoyun,

I am sorry, but I think there is a misunderstanding. The change request:

1. Does not compile - `api/client.go l. 70: Cannot use 'cfg.Client' (type HttpClient) as the type *http.Client`
2. Even if it would compile, it would practically nullify this PR

The PR I issued follows the [Interfaces guidelines](https://github.com/golang/go/wiki/CodeReviewComments#interfaces) to the letter. 
The **producer** package is `net/http` . The **consumer** package is `github.com/prometheus/client_golang/api`.

We need an _HTTP client implementation_ to perform exactly one function:
`Do(req *http.Request) (*http.Response, error)`
and we call it exactly once, inside `func (c *httpClient) Do(ctx context.Context, req *http.Request) (*http.Response, []byte, error)`
(note that both `httpClient` and `httpClient.client` are unexported internal structs).

This is exactly why I have defined the interface in the **consumer** package, and made sure that the native `http.Client` struct complies with this interface (of course). The idea is to enable replacing the native client with implementations which have retries (on rate limiting, network errors etc.), something that the native client does not support (PS this cannot be done in a round tripper, you need a client implementation for that - e.g. `github.com/hashicorp/go-retryablehttp/Client`).

The unexported function `(cfg *Config) client()` you requested to change is called exactly once, from the exported function
`func NewClient(cfg Config) (Client, error)`
If we change the signature back to return the concrete type (even if it would compile), we render this PR meaningless as it will mean you CANNOT replace the native client with another implementation, which was the motivation for this entire PR.

Thank you,
Roy

Reply:
See this example code to use with native http.Client and custom HTTP Client implementations:

```
package main

import (
	"fmt"
	rhttp "github.com/hashicorp/go-retryablehttp"
	"github.com/prometheus/client_golang/api"
	"net/http"
)

type retryClient struct {
	c *rhttp.Client
}

func (rc *retryClient) Do(req *http.Request) (*http.Response, error) {
	if r, err := rhttp.FromRequest(req); err == nil {
		return rc.c.Do(r)
	} else {
		return nil, err
	}
}

func newRetryClient(rt http.RoundTripper) *retryClient {
	rc := rhttp.NewClient()
	rc.HTTPClient = &http.Client{Transport: rt}
	return &retryClient{c: rc}
}

const (
	address = "http://localhost:9090"
)

func main() {
	cfg1 := api.Config{Address: address, RoundTripper: http.DefaultTransport}
	cfg2 := api.Config{Address: address, Client: http.DefaultClient}
	cfg3 := api.Config{Address: address, Client: newRetryClient(http.DefaultTransport)}
	var cfgs = []api.Config{cfg1, cfg2, cfg3}
	for _, cfg := range cfgs {
		c, err := api.NewClient(cfg)
		fmt.Printf("client: %v, err: %v\n", c, err)
		// ...
	}
}
```